NebulaHost/PacketProcessors/GameHistory/GameHistoryTechRefundProcessor.cs

```diff
@@ -0,0 +1,28 @@
+﻿using NebulaModel.Attributes;
+using NebulaModel.Networking;
+using NebulaModel.Packets.GameHistory;
+using NebulaModel.Packets.Processors;
+using System.Collections.Generic;
+using NebulaModel.Logger;
+
+namespace NebulaHost.PacketProcessors.GameHistory
+{
+    [RegisterPacketProcessor]
+    class GameHistoryTechRefundProcessor : IPacketProcessor<GameHistoryTechRefundPacket>
+    {
+        public void ProcessPacket(GameHistoryTechRefundPacket packet, NebulaConnection conn)
+        {
+            //only refund if we have contributed
+            if(packet.TechHashedContributed > 0)
+            {
+                //client should have the same research queued, seek currently needed itemIds and re-add points that were contributed
+                foreach (KeyValuePair<int, int> item in GameMain.data.mainPlayer.mecha.lab.itemPoints.items)
+                {
+                    GameMain.data.mainPlayer.mecha.lab.itemPoints.Alter(item.Key, (int)packet.TechHashedContributed * 3600);
```

Comment:
Can you replace it please with single reference:
```cs
ItemPack itemPoints = GameMain.data.mainPlayer.mecha.lab.itemPoints;
```

and then have it in the loop:
```cs
foreach (KeyValuePair<int, int> item in itemPoints)
...
itemPoints.Alter(item.Key, (int)packet.TechHashedContributed * 3600);
```

Reply:
Good catch :)